### PR TITLE
Add Delphine PCB

### PR DESCRIPTION
**Pull request by mechlovin**
## Description

Add Delphine PCB, Dolpad compatible.
2 version: mono LED and RGB LED.
Thanks!
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/mechlovin/delphine/config.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x4D4C
+#define MANUFACTURER    Mechlovin
+#define PRODUCT         Mechlovin Delphine
+#define DESCRIPTION     Mechlovin Delphine
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/config.h:26]:
```suggestion
```

File: keyboards/mechlovin/delphine/delphine.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    return process_record_user(keycode, record);
+}
+
+bool led_update_kb(led_t led_state) {
+    // put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+    return led_update_user(led_state);
+}
+*/
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/delphine.c:50]:
```suggestion
```

File: keyboards/mechlovin/delphine/delphine.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT_all( \
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/delphine.h:21]:
```suggestion
#define LAYOUT_ortho_6x4( \
```

File: keyboards/mechlovin/delphine/delphine.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    { K00,   K01,   K02,   K03   }, \
+    { K10,   K11,   K12,   K13   }, \
+    { K20,   K21,   K22,   K23   }, \
+    { K30,   K31,   K32,   K33   }, \
+    { K40,   K41,   K42,   K43   }, \
+    { K50,   K51,   K52,   K53   }, \
+}
+
+// generated by KBFirmware JSON to QMK Parser
+// https://noroadsleft.github.io/kbf_qmk_converter/
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/delphine.h:38]:
```suggestion
```

File: keyboards/mechlovin/delphine/info.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+    "keyboard_name": "Delphine",
+    "url": "",
+    "maintainer": "Team Mechlovin'",
+    "width": 4,
+    "height": 6.25,
+    "layouts": {
+        "LAYOUT_all": {
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/info.json:8]:
```suggestion
        "LAYOUT_ortho_6x4": {
```

File: keyboards/mechlovin/delphine/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_all(
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/keymaps/default/keymap.c:19]:
```suggestion
    [0] = LAYOUT_ortho_6x4(
```

File: keyboards/mechlovin/delphine/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_all(
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/keymaps/via/keymap.c:19]:
```suggestion
    [0] = LAYOUT_ortho_6x4(
```

File: keyboards/mechlovin/delphine/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        KC_ESC,  BL_STEP, RGB_TOG, RGB_MOD,
+        KC_NLCK, KC_PSLS, KC_PAST, KC_PMNS,
+        KC_P7,   KC_P8,   KC_P9,   KC_PPLS,
+        KC_P4,   KC_P5,   KC_P6,   KC_PEQL,
+        KC_P1,   KC_P2,   KC_P3,   KC_PENT,
+        KC_P0,   KC_P0,   KC_PDOT, KC_BSPC
+    ),
+    [1] = LAYOUT_all(
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/keymaps/via/keymap.c:27]:
```suggestion
    [1] = LAYOUT_ortho_6x4(
```

File: keyboards/mechlovin/delphine/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    ),
+    [2] = LAYOUT_all(
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/keymaps/via/keymap.c:35]:
```suggestion
    [2] = LAYOUT_ortho_6x4(
```

File: keyboards/mechlovin/delphine/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    ),
+    [3] = LAYOUT_all(
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/keymaps/via/keymap.c:43]:
```suggestion
    [3] = LAYOUT_ortho_6x4(
```

File: keyboards/mechlovin/delphine/rgb_led/rules.mk [hunk 11]
```diff
@@ -0,0 +1,3 @@
+BOOTMAGIC_ENABLE = lite            # Virtual DIP switch configuration
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/rgb_led/rules.mk:1]:
```suggestion
```

File: keyboards/mechlovin/delphine/rules.mk [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+	CONSOLE_ENABLE = no          # Console for debug
+	COMMAND_ENABLE = no          # Commands for debug and configuration
+	# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+	SLEEP_LED_ENABLE = no        # Breathing sleep LED during USB suspend
+	# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+	NKRO_ENABLE = yes            # USB Nkey Rollover
+	BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality
+	RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow
+	BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+	AUDIO_ENABLE = no            # Audio output on port C6
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/rules.mk:22]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

File: keyboards/mechlovin/delphine/rules.mk [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+DEFAULT_FOLDER = mechlovin/delphine/mono_led
```
Comment (zvecr) [inline keyboards/mechlovin/delphine/rules.mk:24]:
```suggestion
LAYOUTS = ortho_6x4
DEFAULT_FOLDER = mechlovin/delphine/mono_led
```

File: keyboards/mechlovin/delphine/info.json [hunk 14]
```diff
@@ -0,0 +1,39 @@
+{
+    "keyboard_name": "Delphine",
+    "url": "",
+    "maintainer": "Team Mechlovin'",
+    "width": 4,
+    "height": 6.25,
+    "layouts": {
+        "LAYOUT_ortho_6x4": {
+            "key_count": 24,
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/info.json:9]:
```suggestion
```

File: keyboards/mechlovin/delphine/rules.mk [hunk 15]
```diff
@@ -0,0 +1,23 @@
+# MCU name
+MCU = atmega32u4
+# Bootloader selection
+BOOTLOADER = atmel-dfu
+# Build Options
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rules.mk:5]:
```suggestion

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+  4,          0,          1,          1,          4,
+} };
+
+void rgb_matrix_indicators_user(void) {
+        if (host_keyboard_led_state().num_lock) {
+            rgb_matrix_set_color(4, 255, 255, 255);
+        }
+}
+
+uint32_t layer_state_set_user(uint32_t state)
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:84]:
```suggestion
layer_state_t layer_state_set_user(layer_state_t state)
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+{
+  // if on layer 1, turn on L1 LED, otherwise off.
+    if (biton32(state) == 0) {
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:87]:
```suggestion
    if (get_highest_layer(state) == 0) {
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        rgb_matrix_set_color(1, 255, 0, 0);
+    } else {
+        rgb_matrix_set_color(1, 0, 0, 0);
+    }
+  // if on layer 2, turn on L2 LED, otherwise off.
+    if (biton32(state) == 1) {
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:93]:
```suggestion
    if (get_highest_layer(state) == 1) {
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        rgb_matrix_set_color(0, 255, 0, 0);
+    } else {
+        rgb_matrix_set_color(0, 0, 0, 0);
+    }
+
+  // if on layer 3, turn on L3 LED, otherwise off.
+    if (biton32(state) == 2) {
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:100]:
```suggestion
    if (get_highest_layer(state) == 2) {
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        rgb_matrix_set_color(3, 255, 0, 0);
+    } else {
+        rgb_matrix_set_color(3, 0, 0, 0);
+    }
+
+  // if on layer 4, turn on L4 LED, otherwise off.
+    if (biton32(state) == 3) {
```
Comment (fauxpark) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:107]:
```suggestion
    if (get_highest_layer(state) == 3) {
```

File: keyboards/mechlovin/delphine/delphine.c [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    writePinLow(B5);
+}
+
+bool led_update_kb(led_t led_state) {
+    if(led_update_user(led_state)) {
+        writePin(B3, led_state.num_lock);
+    }
+
+    return true;
+}
```
Comment (noroadsleft) [inline keyboards/mechlovin/delphine/delphine.c:35]: B5 on the init, but toggling B3?

File: keyboards/mechlovin/delphine/rgb_led/readme.md [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# rgb_led
+
+![rgb_led](imgur.com image replace me!)
+
+A short description of the keyboard/project
+
+* Keyboard Maintainer: [Team Mechlovin'](https://github.com/yourusername)
+* Hardware Supported: The PCBs, controllers supported
+* Hardware Availability: Links to where you can find this hardware
```
Comment (noroadsleft) [inline keyboards/mechlovin/delphine/rgb_led/readme.md:9]: This should be filled out, or the file deleted. (You already have a readme at `keyboards/mechlovin/delphine/readme.md`.)

File: keyboards/mechlovin/delphine/delphine.c (hunk 21 above)
Comment (mechlovin) [reply keyboards/mechlovin/delphine/delphine.c:35]: oh, B5.
Thanks for input.

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c (hunk 16 above)
Comment (drashna) [reply keyboards/mechlovin/delphine/rgb_led/rgb_led.c:84]: And really, it should be 
```suggestion
layer_state_t layer_state_set_kb(layer_state_t state)
```
Or
```suggestion
__attribute__((weak))
layer_state_t layer_state_set_user(layer_state_t state)
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+}, {
+  4,          4,          4,          4,
+  4,          1,          1,          4,
+  4,          1,          1,          4,
+  4,          1,          1,          4,
+  4,          1,          1,          4,
+  4,          0,          1,          1,          4,
+} };
+
+void rgb_matrix_indicators_user(void) {
```
Comment (drashna) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:78]:
```suggestion
void rgb_matrix_indicators_kb(void) {
```
or
```suggestion
__attribute__((weak))
void rgb_matrix_indicators_user(void) {
```

File: keyboards/mechlovin/delphine/rgb_led/rgb_led.c [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+    return state;
+}
+
+#endif
+
+#ifndef RAW_ENABLE
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+#else
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+#endif
```
Comment (drashna) [inline keyboards/mechlovin/delphine/rgb_led/rgb_led.c:122]: Why?

Comment (mechlovin) [reply keyboards/mechlovin/delphine/rgb_led/rgb_led.c:122]: For define RGB matrix

Comment (drashna) [reply keyboards/mechlovin/delphine/rgb_led/rgb_led.c:122]: Let me rephrase then. raw HID, nor via takes over the `process_record_kb` function, so that should be usable here still. Unless I'm missing something, this ... isn't really necessary.

Comment (mechlovin) [reply keyboards/mechlovin/delphine/rgb_led/rgb_led.c:122]: So, it should be 
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
  if (record->event.pressed) {
is it right?

Comment (drashna) [reply keyboards/mechlovin/delphine/rgb_led/rgb_led.c:122]: Yup. But if you want a better way to handle that:
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {

    if (!process_record_user(keycode, record)) { return false; }
    
    if (record->event.pressed) {
```

This isn't standard, but ... I think it may be a better way to implement this, this way.

**Comment by noroadsleft** [comment]
Thanks!
